src/pool/cli.c

```diff
@@ -1535,8 +1560,6 @@ pool_evict_cp(tse_task_t *task, void *data)
 
 	rc = out->pvo_op.po_rc;
 	if (rc != 0) {
-		if (rc == -DER_NOTREPLICA)
-			rc = -DER_NONEXIST;
```

Comment:
so if the selected srv is not raft replica srv, here just return DER_NOTREPLICA to client?
or should it remove the tgt rank from svc list and rechoose?

Reply:
With this patch, such a -DER_NOTREPLICA will cause the rank to be removed from rsvc_client.sc_ranks and will return RECHOOSE. (See rsvc_client_complete_rpc.) Then, when we attempt to rechoose, if there's no rank left in rsvc_client.sc_ranks, -DER_NOTREPLICA will be returned. (I initially return -DER_NONEXIST, but for many cases its meaning may be misleading.)

Reply:
It looks like the 2 client cases that change a -DER_NOTREPLICA returned by rsvc_client_choose() to -DER_NONEXIST are pool evict and pool connect. Is there any reason to keep those two cases or just have them return -DER_NOTREPLICA without any translation?

Reply:
Mostly for meeting the expectation of the existing tests: one in daos_test -p (daos_pool_connect) and one in the Python suite (daos_pool_evict). For the former, I think we'll look up the pool UUID in the pool directory and get a -DER_NONEXIST there, instead of relying on the -DER_NOTREPLICA convert like today. For the latter, I'm not sure what should we do.

Reply:
Any suggestions?

Reply:
I need more time to think through many different scenarios and the current behavior, the behavior with this patch, and how we might like it to be ultimately. Perhaps future work item DAOS-1742 is where this could take place.

Generally, some ideas I have started to consider:
- could we arrange for client API to never return -DER_NOREPLICA to a client
- could we arrange for client API to return a different error (maybe -DER_INVAL?)  to the client if it is discovered in the completion/rechoose/retry logic that one of the pool service replica ranks is out of bounds (got the CaRT OOG error) / versus what would be returned if a list of valid ranks are provided but are not replicas for the given pool resource (-DER_NONEXIST?).
This gets messy/complicated depending on the combination of pool service replica ranks provided by the client (especially if it is a combination of invalid/OOG and valid ranks).

Reply:
Thanks, @kccain. We shall remove the svc parameters from as many API methods as possible, querying the MS's pool directory internally instead. (I didn't find the JIRA ticket.) "This gets messy/complicated"---I had exactly the same thought. I think the simplest solution is to report individual rank issues in the log, while keeping trying other ranks until there's no more ranks to try (at which point we shall query the MS for any updated replica list in the future).